Basic vue UI

Henny022 opened the pull request:
This is based on the new backend by @knusperkrone

Henny022 committed:
added base vue app

Henny022 committed:
added bootstrap

Henny022 committed:
first mock pages

Henny022 committed:
login and (very simple) photo viewing

Henny022 committed:
uploading of photos

Henny022 committed:
modal component

Henny022 committed:
upload modal

Henny022 committed:
Photo Mock done

Henny022 committed:
user mockup
added some more structure to the files

Henny022 committed:
typescript object for API interaction

Henny022 committed:
Photo API

Henny022 committed:
User Api

Henny022 committed:
code cleanup
renamed views to avoid conflicts with api objects

georg-schwarz commented on ui/src/ApiThing.ts:
```diff
@@ -0,0 +1,139 @@
+  email: string;
+}
+
+class ApiThing {
```
You can create singletons in js by
```javascript
const apiThing = new ApiThing();
export apiThing;
```

This lets you just `import {apiThing} from './.../ApiThing'` and you don't have to pass it through all components!

georg-schwarz replied:
`ApiThing` is not a good name, and does too many things at once. I'd split up in `AuthApi`, `PhotoApi` and `UserApi`

Henny022 replied:
I agree on the name, I already renamed it to WahlzeitApi, just didn't commit that yet.
I made it a single class, since I'd say it only does a single thing, that being handling the interaction with the backend API (one class for one API).
Splitting that would require some base class that handles the basic communication that would be shared by all the individual API's.

I think that any code using a singleton is bad code for multiple reasons, but if it is the preferred solution in your project I can change that.
If I do split the the API class, that would make 3+ singletons.

georg-schwarz replied:
Well, I agree that the singleton approach is not optimal as well. In general, dependency injection is the way to go in my opinion. Using the props is a way to inject dependencies... However, I'd still go with the instance export way because it makes the component interface smaller. But maybe I'm wrong - did you look into best practices for vue regarding this?

Single-Responsibility is always a matter of perspective... Since the API is still slim, lets leave it as you designed it.

Henny022 replied:
Vue has provide/inject for dependency injections, but that has some limitations and since I use the api in almost every component anyways I chose props. The documentation also isn't very clear on how to use it with Vue 3 and TS class components, and due to time constraints I couldn't do a lot of testing.
There are some plugins that would help here, but I didn't want to add those as additional dependencies.
I might just change it to a singleton for reading simplicity.

For the Single-Responsibility I just think that adding multiple classes with only 2-3 methods each would be overkill.

georg-schwarz commented on ui/src/views/PhotoView.vue:
```diff
@@ -0,0 +1,59 @@
+
+  mounted() {
+    this.api
+      ?.getPhoto(this.id)
```
Pls use the `aync/await` syntax for better code readablity: typescriptlang.org/docs/handbook/release-notes/typescript-1-7.html

georg-schwarz replied:
Same for other views

Henny022 replied:
Not sure I understand what you mean.
So replace the use of promises with await?
The link you posted just leads me to typescript.org

georg-schwarz replied:
Yes, those promises look like spaghetti code, so please use the `async/await` syntax instead.

Henny022 committed:
cleaned Api interaction
renamed class
using singleton to reduce used props
using await instead of promises

Henny022 committed:
started adding mocks for photo case related stuff

Henny022 committed:
small fixes

Henny022 committed:
cleaned photo actions

Henny022 committed:
Added some stuff along the way
Added signup modal
Added search functionality (currently broken)
cleaned up the codebase

Henny022 committed:
fixed faulty search form

georg-schwarz commented:
Hey Hendik, I just checked out your PR and saw a few issues in my editor:

(1) Please rebase your PR on the `wahlzeit-40-new-ui` branch
(2) Some template roots have multiple elements (only one allowed).
(3) Your alias-based imports (@) are shown as error. Is there something required to be configured locally?
(4) The Vetur plugin shows another error: Property '$router' does not exist on type 'App' - probably something is not configured in the shim?
(5) Please add the UI to the Docker setup. You can find a dockerfile for vue here: https://github.com/jvalue/open-data-service/blob/main/ui/Dockerfile
